### ISSUE TITLE
zcompile

### ISSUE THREAD

**Issue by mattmc3** [mattmc3/antidote#53]
zcompile cloned plugins by default.

Add a zstyle to disable.

**Comment by peikk0** [comment]
zcompiling the generated static file would be great too.

**Comment by GreyXor** [comment]
@mattmc3 Hello, i'm interested in this feature, it's hard to implement ?

**Comment by mattmc3** [comment]
It's not hard to implement. In fact, I think I can cut a 1.9.4 release with this feature pretty quickly. However, I no longer believe that having this be the default behavior is ideal. I'd rather have this be an opt-in feature since all sorts of weird scenarios can happen if people don't know what they're getting into with `zcompile`. To opt-in to this feature, I'll add the following zstyles:

```zsh
# zcompile all bundles
zstyle ':antidote:bundle:*' zcompile 'yes'

# zcompile the static file
zstyle ':antidote:static' zcompile 'yes'
```

**Comment by GreyXor** [comment]
Awesome, I saw your new branch, [v1.9.4](https://github.com/mattmc3/antidote/tree/v1.9.4), it's already working? I can give a try if you want

**Comment by mattmc3** [comment]
Thanks @GreyXor. I think that branch is ready to test. Let me know how it works for you. Don't forget those zstyles. For completeness, the things I have tried to verify or added unit tests to ensure:
- antidote bundle will clone and zcompile repos
- antidote load will zcompile the static file
- antidote update will recompile .zwc files
- setting `zstyle ':antidote:bundle:ohmyzsh/ohmyzsh' zcompile 'no'` will ensure that ohmyzsh/ohmyzsh is excluded from compilation (just as an example, not because you need to).
- test-data folders are excluded from compilation, such as those that come with [zsh-users/zsh-syntax-highlighting](https://github.com/zsh-users/zsh-syntax-highlighting/tree/master/highlighters/brackets/test-data)

Let me know if you find any other scenarios that we need to account for.

**Comment by GreyXor** [comment]
THanks mattmc3, it's a great job!
I'm redoing my zsh configuration from scratch so I'll probably test a lot of things without forgetting zcompile :+1: I'll let you know how my tests going

**Comment by GreyXor** [comment]
There's a way to also compile local plugin?
Example I have this `.zsh_plugins.txt`
```bash
.local-plugin
```
.local-plugin contain a zsh script
 I would like antidote also zcompile my local plugins

**Comment by mattmc3** [comment]
Hmm. How would antidote know that .local-plugin is a proper zsh file it can/should zcompile? If it is named .local-plugin.zsh, that might make it easier. There may be a simple way I'm not immediately thinking of.

**Comment by mattmc3** [comment]
I think I have a solution and just pushed a commit with some tests that verify this use case (no extension) should now be covered. Let me know.

**Comment by GreyXor** [comment]
@mattmc3 Yes that's working as expected.

Do you think you can also make zcompile to the `zsh_plugins.zsh` file ?

That's my `.zshrc`

```zsh
# Recompile/Setup everything if `zsh_plugins.zsh` is new than `.zsh_plugins.txt` or if .cache/antidote is older than one week.
if [[ ! "$HOME/.cache/antidote/zsh_plugins.zsh" -nt "$HOME/.zsh_plugins.txt" || $(($(date +%s) - $(stat -c %Y "$HOME/.cache/antidote"))) -gt 604800 ]]; then
  (
    rm -rf .cache/antidote
    mkdir -p .cache/antidote
    source /usr/share/zsh-antidote/antidote.zsh
    zstyle ':antidote:bundle:*' zcompile 'yes'
    zstyle ':antidote:static' zcompile 'yes'
    antidote bundle < "$HOME/.zsh_plugins.txt" > "$HOME/.cache/antidote/zsh_plugins.zsh"

    for file in .zshrc .zshenv .zprofile "$HOME/.cache/antidote/zsh_plugins.zsh"; do
      zcompile "$file"
    done
  )
fi
source "$HOME/.cache/antidote/zsh_plugins.zsh"
```

As you can see I manually zcompile the `zsh_plugins.zsh`, maybe antidote can handle this automatically ?

and the rest is working fine so far. I'm still testing

**Comment by GreyXor** [comment]
```zsh
# Reinstall/setup everything if .zsh newer than .txt or if .cache/antidote is older than 1 week.
if [[ ! "$HOME/.cache/antidote/zsh_plugins.zsh" -nt "$HOME/.zsh_plugins.txt" || $(($(date +%s) - $(stat -c %Y "$HOME/.cache/antidote"))) -gt 604800 ]]; then
  (
    rm -rf .cache/antidote
    mkdir -p .cache/antidote
    source /usr/share/zsh-antidote/antidote.zsh
    zstyle ':antidote:bundle:*' zcompile 'yes'
    zstyle ':antidote:static' zcompile 'yes'

    antidote bundle < "$HOME/.zsh_plugins.txt" > "$HOME/.cache/antidote/zsh_plugins.zsh"

    for file in .zshrc .zshenv .zprofile "$HOME/.cache/antidote/zsh_plugins.zsh"; do
      zcompile "$file"
    done
  )
fi
source "$HOME/.cache/antidote/zsh_plugins.zsh"
```

Just sharing my  `.zshrc`, everything is working as expected, continuing my tests :+1:

**Comment by mattmc3** [comment]
@GreyXor  - awesome news! I pushed one more commit - this one ensures your static file is compiled by emitting the Zsh script to do so on the `antidote bundle` command. I think this should be the last thing needed to complete the feature. Any additional feedback welcome before I cut a release.

**Comment by GreyXor** [comment]
@mattmc3 - I've tested it and I think everything's fine. I had no bugs during my tests and everything got zcompiled.

**Comment by devnoname120** [comment]
@mattmc3 Could you expand on this:
> all sorts of weird scenarios can happen if people don't know what they're getting into with `zcompile`?

I'm not very familiar with `zcompile` and I would like to know what I'm getting myself into 😄

**Comment by mattmc3** [comment]
You can read more here: https://www.zsh.org/mla/users/2021/msg01080.html

TLDR; zcompile is a form of caching, and you can get weird behavior if your cache is invalid. There's trade-offs, so it's only a callout that while antidote supports zcompile, I can't provide troubleshooting of any side-effects of using that feature.
